### PR TITLE
Fix ecrecover memory clear

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 1]
```diff
@@ -8223,6 +8223,32 @@ BOOST_AUTO_TEST_CASE(failing_ecrecover_invalid_input)
...
+					0x6944c77849b18048f6abe0db8084b0d0d0689cdddb53d2671c36967b58691ad4,
+					0xef4f06ba4f78319baafd0424365777241af4dfd3da840471b4b4b087b7750d0d,
+					0xca35b7d915458ef540ade6068dfe2f44e8fa733c,
+					0xca35b7d915458ef540ade6068dfe2f44e8fa733c
+				);
+			}
+			function recover(bytes32 hash, uint8 v, bytes32 r, bytes32 s, uint blockExpired, bytes32 salt)
+				returns (address)
+			{
+				require(hash == sha3(blockExpired, salt));
```
Comment (axic) [inline test/libsolidity/SolidityEndToEndTest.cpp]: What is the point having `blockExpired`/`salt`?

Comment (chriseth) [reply test/libsolidity/SolidityEndToEndTest.cpp]: It was just part of the example code that triggered the bug.

Comment (chriseth) [reply test/libsolidity/SolidityEndToEndTest.cpp]: Oh, now I see. No, we actually need it, computing `sha3(blockExpired, salt)` is the part that puts the correct data into the correct point in memory that was not properly zeroed out.

Comment (axic) [reply test/libsolidity/SolidityEndToEndTest.cpp]: In that case wouldn’t an assembly piece to write random stuff at the free memory pointer would be the most definitive test? This test now depends on sha3 behaving as it is. (Also we renamed it to keccak256 internally :) )

**Comment by chriseth** [comment]
Added the assembly test.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 2]
```diff
@@ -8251,6 +8251,53 @@ BOOST_AUTO_TEST_CASE(failing_ecrecover_invalid_input)
...
```
Comment (axic) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Please use `keccak256` instead of `sha3` as in the rest of the tests.

File: docs/bugs.json [hunk 3]
```diff
@@ -1,5 +1,12 @@
 [
     {
+        "name": "ECRecoverMalformedInput",
+	"summary": "The ecrecover() builtin can return garbage for malformed input.",
```
Comment (axic) [inline docs/bugs.json:4]: Just marking it here: this indentation is off and below too. We'll fix later.

**Review by axic — APPROVED** [review]

**Pull request merged**